coveo-functools/coveo_functools/flex/deserializer.py

```diff
@@ -14,113 +11,38 @@
     get_origin,
     Union,
     Dict,
-    Callable,
     Tuple,
     overload,
-    Final,
     List,
     Sequence,
     cast,
     Iterable,
+    Callable,
 )
 
 from coveo_functools.annotations import find_annotations
-from coveo_functools.casing import unflex, flexcase, TRANSLATION_TABLE
+from coveo_functools.casing import TRANSLATION_TABLE, unflex
 from coveo_functools.dispatch import dispatch
 from coveo_functools.exceptions import UnsupportedAnnotation
+from coveo_functools.flex.types import TypeHint, PASSTHROUGH_TYPES
 
-_ = unflex, flexcase  # mark them as used (forward compatibility vs docs)
-
-
-JSON_TYPES = (
-    str,
-    bool,
-    int,
-    float,
-    type(None),
-    dict,
-)  # list omitted to support list of custom types
-PASSTHROUGH_TYPES = {None, Any, *JSON_TYPES}
 
 T = TypeVar("T")
 
-TypeHint = Any  # :shrug:
 
-RealClass = Type[T]
-RealFunction = Callable[..., T]
-RealObject = Union[RealClass, RealFunction]
-WrappedClass = Type[T]
-WrappedFunction = Callable[..., T]
-WrappedObject = Union[WrappedClass, WrappedFunction]
-
-
-RAW_KEY: Final[str] = "_coveo_functools_flexed_from_"
-
-
-@overload
-def flex() -> Callable[[RealObject], WrappedObject]:
-    ...
-
-
-@overload
-def flex(obj: None) -> Callable[[RealObject], WrappedObject]:
-    ...
-
-
-@overload
-def flex(obj: RealClass) -> WrappedClass:
-    ...
-
-
-@overload
-def flex(obj: RealFunction) -> WrappedFunction:
-    ...
-
-
-def flex(
-    obj: Optional[RealObject] = None,
-) -> Union[WrappedObject, Callable[[RealObject], WrappedObject]]:
-    """Wraps `obj` into recursive flexcase magic."""
-    if obj is not None:
-
-        """
-        Covers decorator usages without parenthesis:
-
-            @flex
-            def obj(...)
-
-            @flex
-            class C:
-                @flex
-                def obj(self, ...)
-                    ...
-
-        Also covers the complete inline usage:
-
-            f = flex(obj, ...)(**dirty_kwargs)
-
-        """
-
-        return _generate_wrapper(obj)
-
-    else:
-
-        """
-        Covers decorator usages with parenthesis:
-
-            @flex()
-            def obj(...)
-
-            @flex()
-            class C:
+def prepare_payload_for_unpacking(
+    fn: Callable[..., T], dirty_kwargs: Dict[str, Any]
+) -> Dict[str, Any]:
+    """Return a copy of `dirty_kwargs` that can be `**unpacked` to fn()."""
+    mapped_kwargs = unflex(fn, dirty_kwargs)
+    converted_kwargs = {}
 
-                @flex()
-                def obj(self, ...)
-                    ...
-        """
+    for arg_name, arg_type in find_annotations(fn).items():
+        if arg_name not in mapped_kwargs:
+            continue  # this may be ok, for instance if the target argument has a default
+        converted_kwargs[arg_name] = deserialize(mapped_kwargs[arg_name], hint=arg_type)

```

Comment:
This is not new, it was moved from the end of this file. It was called `_remap_and_convert`.